bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/ldb/ReadAheadManager.java

```diff
@@ -0,0 +1,655 @@
+    }
+
+    /**
+     * A structure that records the transitions of the task status.
+     */
+    public static class ReadAheadTaskStatus {
+        private long ledgerId;
+        private long startEntryId;
+        private long endEntryId = -1;
+        private long readAheadEntries;
+
+        private final long readAheadTaskExpiredTimeMs;
+        private final long readAheadTimeoutMs;
+        private long readCompletedTimeMs = -1;
+
+        private final ReentrantLock lock = new ReentrantLock();
+        private final Condition condition = lock.newCondition();
+        private volatile boolean readCompleted = false;
+
+        public ReadAheadTaskStatus(long ledgerId, long startEntryId, int readAheadEntries,
+                                   long readAheadTaskExpiredTimeMs, long readAheadTimeoutMs) {
+            this.ledgerId = ledgerId;
+            this.startEntryId = startEntryId;
+            this.readAheadEntries = readAheadEntries;
+
+            this.readAheadTaskExpiredTimeMs = readAheadTaskExpiredTimeMs;
+            this.readAheadTimeoutMs = readAheadTimeoutMs;
+        }
+
+        public boolean hasEntry(long entryId) {
+            if (!readCompleted) {
+                // the max-estimate is return before the read-ahead task is completed
+                return entryId >= startEntryId && entryId <= startEntryId + readAheadEntries;
+            } else {
+                return entryId >= startEntryId && entryId <= endEntryId;
+            }
+        }
+
+        public void updateEndEntry(long endEntryId) {
+            this.endEntryId = endEntryId;
+        }
+
+        public long getEndEntryId() {
+            return endEntryId;
+        }
+
+        public boolean isExpired() {
+            return readCompletedTimeMs > 0
+                    && (System.currentTimeMillis() - readCompletedTimeMs) > readAheadTaskExpiredTimeMs;
+        }
+
+        public void waitUntilReadCompleted(DbLedgerStorageStats dbLedgerStorageStats) {
+            long blockStartNanos = MathUtils.nowInNano();
+            lock.lock();
+            try {
+                while (!readCompleted) {
+                    try {
+                        condition.await(readAheadTimeoutMs, TimeUnit.MILLISECONDS);
+                    } catch (InterruptedException e) {
+                        log.warn("Failed to read entries={} ahead from L{} E{} due to timeout={}ms",
+                                readAheadEntries, ledgerId, startEntryId, readAheadTimeoutMs);
+                        readCompleted();
+                    }
+                }
+            } finally {
+                lock.unlock();
+
+                // record request blocking time
+                recordStatsInNano(dbLedgerStorageStats.getReadAheadAsyncBlockTime(), blockStartNanos);
+            }
+        }
+
+        public void readCompleted() {
+            lock.lock();
+            try {
+                readCompleted = true;
+                condition.signalAll();
```

Comment:
This is not the case, the `signalAll` here will only wake up all threads that are waiting for this read-ahead data, and those threads waiting for other read-ahead data will not be affected. Because the minimum granularity of read-ahead locks is an independent read-ahead task.